KAFKA-5702: extract refactor StreamThread

dguy opened the pull request:
Extracted `TaskManager` to handle all task related activities.
Make `StandbyTaskCreator`, `TaskCreator`, and `RebalanceListener` static classes so they must define their dependencies and can be testing independently of `StreamThread`
Added interfaces between `StreamPartitionAssignor` & `StreamThread` to reduce coupling.

dguy committed:
extract taskmanager from stream thread
make partition assignor test work

dguy commented:
@guozhangwang, @bbejeck, @enothereska, @mjsax, @sriramsub - Sorry about the size of this change, but it is pretty difficult to pull it apart without it being so big. I don't think it is by any means perfect or complete, but I hope it can put us on a better path. There is way too much coupling in the internals stuff at the moment and we really need to think about defining and using appropriate interfaces.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:
```diff
@@ -62,6 +63,18 @@
     private String logPrefix;
 
+    // interface to get info about the StreamThread
+    interface ThreadDataProvider {
```
Perhaps separate class file?

guozhangwang replied:
+1

enothereska commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/Task.java:
```diff
@@ -0,0 +1,67 @@
+import java.util.Map;
+import java.util.Set;
+
+public interface Task {
```
Ideally has more comments.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskIdToPartitionsProvider.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.Map;
+import java.util.Set;
+
+public interface TaskIdToPartitionsProvider {
```
Not immediately clear what this does, a comment would help.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6576/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6561/
Test PASSed (JDK 8 and Scala 2.12).

bbejeck commented:
@dguy thanks for taking this on.  

My thoughts after a first pass is can we extract  the inner-classes (`RebalanceListener`,`TaskCreator` and `StandbyTaskCreator` etc) from `StreamThread`? Although I'm pretty sure those classes are only used by `StreamThread` IMHO it's worth the trade-off to reduce `StreamThread` by a few hundred LOC.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -144,8 +162,18 @@ public void close(final boolean clean) {
 
-    Map<TopicPartition, Long> checkpointedOffsets() {
+    @Override
+    public int addRecords(final TopicPartition partition, final Iterable<ConsumerRecord<byte[], byte[]>> records) {
```
I think `StandbyTask#addRecords` will never be called, so it is OK to be more strict and throw unsupported exception here?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -133,6 +133,24 @@ public void close(final boolean clean) {
     }
 
+    @Override
+    public void closeSuspended(final boolean clean, final RuntimeException e) { }
```
Ditto as below, for these four functions could we be more strict and not calling it in `StreamThread` and throw unsupported exception if ever called?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:
```diff
@@ -62,6 +63,18 @@
+        UUID processId();
+        StreamsConfig config();
+        PartitionGrouper partitionGrouper();
+        void setPartitionAssignor(final StreamPartitionAssignor assignor);
```
Can we pass in the `TaskIdToPartitionsProvider` interface not the `StreamPartitionAssignor` here? If yes we can rename the function as well.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -388,93 +506,69 @@ void removeAllSensors() {
+    private long lastCommitMs;
     private String originalReset;
     private StreamPartitionAssignor partitionAssignor;
+    private boolean processStandbyRecords = false;
```
Line 523: partitionAssignor seems not needed anymore?

dguy replied:
Actually, we do need it to refresh metadata state on line 240

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:
```diff
@@ -53,7 +54,7 @@
 import static org.apache.kafka.streams.processor.internals.InternalTopicManager.WINDOW_CHANGE_LOG_ADDITIONAL_RETENTION_DEFAULT;
 
-public class StreamPartitionAssignor implements PartitionAssignor, Configurable {
+public class StreamPartitionAssignor implements PartitionAssignor, Configurable, TaskIdToPartitionsProvider {
```
One general comment: current there is a mutual dependency between `StreamPartitionAssignor` and `StreamThread` that each of the object needs to maintain a reference of each other. Making them extending an interface to leak as little functionalities as possible is better but with `TaskManager` we still have a circle dependency or `task manager -> partition assignor -> thread data provider (stream thread) -> task manager`.

I'm wondering if we can break this circular dependency. One idea is that we can break partition assignor's dependency on stream thread:

1. builder / name / processorId / config / partitionGrouper are static and won't change after being constructed. We can pass them once in `configure`  from StreamThread when constructing the consumer, but not passing in the `ThreadDataProvider aka StreamThread` itself.

2. `prevActiveTasks` and `cachedTasks` are from `task manager`, we can let `task manager` to hold a reference of the `assignor` since it is dependent on that anyways (so we can replace `setPartitionAssignor` as well), so that it can update these two fields whenever necessary.

So that the dependency now becomes `stream thread -> task manager -> partition assignor`.

guozhangwang replied:
In addition, in unit tests we should consider mock all stream thread's dependent modules like task manager and task creator, and clients. Similar in partition assignor test we should mock any dependent functionalities like `previous active task` and `cached task`

dguy replied:
@guozhangwang - i'll have a look at the dependencies. I think that sounds good. As for the unit tests, i totally agree, but... I wanted to try and keep this PR to a minimum.

dguy replied:
@guozhangwang i'm not really sure what you mean in point 2 above? Are you suggesting that `TaskManager` pushes `cachedTasks` and `prevActiveTasks` on to `StreamPartitionAssignor`? So the interface between `TaskManager` and `StreamsPartitionAssignor` would also need to have `setCachedTasks(..)` and `setPreviousActivetTasks(...)`

dguy replied:
Hmm even for 1. you still need to pass in some interface so that the `StreamPartitionAssignor` can still be set back on `StreamThread/TaskManager`.
I think i'll leave it as is for now and we can refactor further in a follow up

guozhangwang replied:
I agree that we can leave this in a follow-up PR to keep it minimum. Just leaving more details on my proposal:

1. Remove the `ThreadDataProvider` interface, not let StreamThread to extend it.
2. For all its interface functions, builder / name / processId / config / partitionGrouper can be set in the `configure` call: the thread can set these when constructing the config map for the consumer, which will be passed in the `configure` function; or they can be passed by TaskManager as well (see below)
3. In `StreamThread`, it only needs to call the `activeTasks()` and `standbyTasks()` functions of the assignor. We can move these two functions to the `TaskManager` so that StreamThread would not need to call `setPartitionAssignor` and hold a reference on `assignor` any more, it only needs to call `taskManager.activeTasks()` etc.
4. TaskManager now needs `activeTasks` and `standbyTasks`, and the assignor now needs `prevActiveTasks` and `cachedTasks`. One idea is to extract the task assignment logic from `StreamPartitionAssignor` to `TaskManager`, so that the StreamPartitionAssignor becomes a thin class which only needs to hold a reference of the `PartitionGrouper` and `TaskManager` to call their APIs to construct the map from `partition -> task` and `task -> client-host` to implement the `PartitionAssignor` interface. In this case it does not need `builder, prevActiveTasks, cachedTasks, processId` any more, only `name` for log4j.

WDYT?

dguy replied:
Sounds reasonable

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -436,6 +438,11 @@ void closeSuspended(boolean clean, RuntimeException firstException) {
 
+    @Override
+    public Map<TopicPartition, Long> checkpointedOffsets() {
+        return Collections.emptyMap();
```
Similar here, we can throw unsupported exception and make sure `active task`'s function is never called in thread?

dguy replied:
Sure - i was on the fence with doing that. I'm ok either way

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -286,48 +314,144 @@ void retryWithBackoff(final Map<TaskId, Set<TopicPartition>> tasksToBeCreated, f
+                        time,
+                        createProducer(taskId));
+            } finally {
+                log.trace("{} Created active task {} with assigned partitions {}", logPrefix, taskId, partitions);
```
This log entry could be misleading, that even if there is an exception happened and hence no task created, it will still print as `created ...`; and in practice I have once encountered this issue before which affected the trouble shooting process, I think we should try to piggy-back fix it.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -286,48 +314,144 @@ void retryWithBackoff(final Map<TaskId, Set<TopicPartition>> tasksToBeCreated, f
+                try {
+                    return new StandbyTask(taskId, applicationId, partitions, topology, consumer, storeChangelogReader, config, streamsMetrics, stateDirectory);
+                } finally {
+                    log.trace("{} Created standby task {} with assigned partitions {}", logPrefix, taskId, partitions);
```
Ditto here.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -483,34 +577,101 @@ public StreamThread(final InternalTopologyBuilder builder,
+        taskManager.setConsumer(consumer);
+    }
+
+    @SuppressWarnings("ConstantConditions")
```
Just curious what is this for?

dguy replied:
it is to supress the warning caused by this line:
`final int rebalanceTimeoutMs = (Integer) ConfigDef.parseType(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG, maxPollInterval, Type.INT);`

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -0,0 +1,520 @@
+    private final Time time;
+    private final String logPrefix;
+    private final Consumer<byte[], byte[]> restoreConsumer;
+    private final StreamThread.AbstractTaskCreator taskCreator;
```
nit: maybe we can separate AbstractTaskCreator and its two impls into separate classes once we are finalizing the refactoring in a follow-up PR (this PR can stay as is to keep it from exploding LOC)

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -486,6 +468,23 @@ public boolean conditionMet() {
+        if (eosEnabled) {
+            clientSupplier = new MockClientSupplier(applicationId);
+        }
+        return StreamThread.create(internalTopologyBuilder,
```
nit: in unit test we may want to avoid calling `create` because it will construct a lot of unnecessary modules like task manager / created / clients etc, we should consider mocking all these in `StreamThreadTest`.

dguy replied:
Agreed - just probably in a follow up PR. Trying to minimize the size of the change

guozhangwang reviewed (commented):
The refactoring looks promising to me. I left a general comment about breaking circular dependencies.

guozhangwang commented:
@dguy LMK if you'd like to take a further stab on this PR to break the circular dependencies, otherwise the PR looks good in general to me.

dguy committed:
address comments

dguy commented:
@guozhangwang i've addressed your comments. I've left the circular dependency for now as i think it needs some more thought on how to do it properly. We can address that and moving `TaskCreator` etc in a separate PR - WDYT?

dguy commented:
System test results here: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2017-08-09--001.1502280749--dguy--stream-thread-refactor--4d8960d/report.html

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6660/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6645/
Test FAILed (JDK 8 and Scala 2.12).

guozhangwang commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6666/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6651/
Test PASSed (JDK 8 and Scala 2.12).

guozhangwang reviewed (approved):
Left a more detailed suggestion in https://github.com/apache/kafka/pull/3624#discussion_r132614639.

LGTM for this PR. Please feel free to merge.

dguy commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6686/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6701/
Test PASSed (JDK 7 and Scala 2.11).

dguy commented:
merged to trunk
